core/vm/evm.go

```diff
@@ -220,6 +227,30 @@ func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig
 		chainRules:  chainConfig.Rules(blockCtx.BlockNumber, blockCtx.Time),
 	}
 	evm.interpreter = NewEVMInterpreter(evm)
+
+	// If the predicate results are already available, use them.
+	if blockCtx.PredicateResults != nil {
+		return evm
+	}
+
+	// Parse the predicate results from the extra field and store them in the
+	// block context.
+	predicateBytes := header.PredicateBytesFromExtra(blockCtx.Extra)
+	if len(predicateBytes) == 0 {
+		return evm
+	}
+
+	// The VM has already verified the correctness of the results during header
+	// validation.
+	results, err := predicate.ParseResults(predicateBytes)
+	if err != nil {
+		log.Error("Unexpected error parsing predicate results",
+			"err", err,
+		)
+		return evm
+	}
+
+	evm.Context.PredicateResults = results
```

Comment:
Note: because `BlockContext` is pass-by-value, this does not cache `results` for future calls to `NewEVM`.

Reply:
can we comment that?